extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java

```diff
@@ -158,20 +162,48 @@ void registerNativeImageResources(BuildProducer<ServiceProviderBuildItem> servic
                 .produce(ServiceProviderBuildItem.allProvidersFromClassPath(SmallRyeGraphQLConfigMapping.class.getName()));
     }
 
+    @BuildStep
+    SmallRyeGraphQLIndexBuildItem createIndex(TransformedClassesBuildItem transformedClassesBuildItem) {
+        Map<String, byte[]> modifiedClasses = new HashMap<>();
+        Map<Path, Set<TransformedClassesBuildItem.TransformedClass>> transformedClassesByJar = transformedClassesBuildItem
+                .getTransformedClassesByJar();
+        for (Map.Entry<Path, Set<TransformedClassesBuildItem.TransformedClass>> transformedClassesByJarEntrySet : transformedClassesByJar
+                .entrySet()) {
+
+            Set<TransformedClassesBuildItem.TransformedClass> transformedClasses = transformedClassesByJarEntrySet.getValue();
+            for (TransformedClassesBuildItem.TransformedClass transformedClass : transformedClasses) {
+                modifiedClasses.put(transformedClass.getClassName(), transformedClass.getData());
+            }
+        }
+        return new SmallRyeGraphQLIndexBuildItem(modifiedClasses);
+    }
+
     @Record(ExecutionTime.STATIC_INIT)
     @BuildStep
     void buildExecutionService(
             BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer,
             BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchyProducer,
             BuildProducer<SmallRyeGraphQLInitializedBuildItem> graphQLInitializedProducer,
             SmallRyeGraphQLRecorder recorder,
+            SmallRyeGraphQLIndexBuildItem graphQLIndexBuildItem,
             BeanContainerBuildItem beanContainer,
             CombinedIndexBuildItem combinedIndex,
             SmallRyeGraphQLConfig graphQLConfig) {
 
-        IndexView index = combinedIndex.getIndex();
+        Indexer indexer = new Indexer();
+        Map<String, byte[]> modifiedClases = graphQLIndexBuildItem.getModifiedClases();
+
+        for (Map.Entry<String, byte[]> kv : modifiedClases.entrySet()) {
+            try (ByteArrayInputStream bais = new ByteArrayInputStream(kv.getValue())) {
+                indexer.index(bais);
+            } catch (IOException ex) {
+                LOG.warn("Could not index [" + kv.getKey() + "] - " + ex.getMessage());
+            }
+        }
+
+        CompositeIndex compositeIndex = CompositeIndex.create(indexer.complete(), combinedIndex.getIndex());
```

Comment:
Does this ensure that the modified classes are read from their index and not the original index?

Reply:
Seems like order is important here (that makes sense). The second index does a 'putIfAbsent' of sorts. That is what I think happens at least. Maybe @Ladicek can confirm.

Reply:
Right, let's be sure about it before commiting to this.

Reply:
`CompositeIndex` indeed maintains an ordered collection of indices and queries them in order, but most operations query _all_ underlying indices and merge the results. So for example if your composite index contains the same annotated class twice and you call `CompositeIndex.getAnnotations(...)`, you will get the annotations on that class twice.

There's a couple of short-circuiting methods on `CompositeIndex`, notably `getClassByName`, but most methods behave as I described above.

Reply:
@Ladicek -  So are we safe to proceed with this ?

Reply:
I have no idea? :-)

Reply:
To be more specific -- I don't know how `SchemaBuilder` uses the index, so I don't know if what you're doing here is safe. I can only say that the `putIfAbsent` intuition is mostly _not_ correct.

Reply:
Ok thanks, at the moment schema builder  use 
`getClassByName` - see 
`https://github.com/smallrye/smallrye-graphql/blob/cb603e4e4f5dac1c11d60377ed66a4175d51b9ad/common/schema-builder/src/main/java/io/smallrye/graphql/schema/SchemaBuilder.java#L218`

So that is why this is working I guess. 

The other option is to do what is done in OpenAPI: https://github.com/smallrye/smallrye-open-api/blob/main/core/src/main/java/io/smallrye/openapi/runtime/scanner/FilteredIndexView.java

w.d.y.t ?

Reply:
I also see https://github.com/smallrye/smallrye-graphql/blob/cb603e4e4f5dac1c11d60377ed66a4175d51b9ad/common/schema-builder/src/main/java/io/smallrye/graphql/schema/SchemaBuilder.java#L105-L106 and a bunch of other similar calls, which can possibly return one annotation several times. I _guess_ this entire thing works because the "supplemental" index here never contains such classes.

Seems a little fragile to me, because you never know what `TransformedClassesBuildItem` may contain. For example, if a class annotated `org.eclipse.microprofile.graphql.GraphQLApi` uses `io.quarkus.logging.Log`, which is implemented using bytecode transformation, would `TransformedClassesBuildItem` also contain that class? If so, you've got a problem.

The _proper_ solution is what ArC does -- build a metadata overlay on top of Jandex. I briefly thought perhaps some generic overlay thing could be added to Jandex itself, but I don't really know what that would entail.

Reply:
Does ArC do the same as what I pointer to in OpenAPI above ?

Reply:
Not really, no.

I thought a little more about this and my opinion is this. What `CompositeIndex` allows is to take 2 indices that don't overlap, and present them as 1. When there's overlap, `CompositeIndex` is most likely a bad choice.

What you need here, IIUC, is to take an index, and overlay another index on top of it. There's nothing like that in Jandex, AFAIK. And some operations on the `IndexView` interface seem rather hard to implement in an overlay fashion.

If you had your own metadata layer (i.e., your own representation of classes, methods, annotations, etc.), then that layer would be natural for implementing an overlay. If you use Jandex directly, there's no natural place where to put the overlay. (ArC is somewhere in the middle. It uses Jandex types directly, but it always reads class/method/field annotations from an overlay layer, called `AnnotationStore`. That is enough for ArC, but not sure if something like that would be enough for you.)

I guess maybe this strategy could work: admit that there are 2 indices. Read the top-level metadata (e.g. classes annotated `@GraphQLApi`) from the primary index. From there, you'll also find the "entity" classes (or what's the proper term). When reading details about them, try the secondary index first. Something like that.

Reply:
@Ladicek - What would be cool (if possible) in Jandex, is similar to `CompositeIndex` but a `OverrideIndex` that does the `putIfAbsent`. Not sure how easy that would be. For now I am going to add a Index that does that (similar to the OpenAPI Filter)

Reply:
Yes I agree. I _think_ some methods on `IndexView` are hard to implement, or perhaps rather hard to _define_, but the most common ones should be possible.

Reply:
Filed https://github.com/wildfly/jandex/issues/142

Reply:
Thanks @Ladicek . Please have a look at my latest attempt. This works for me, and I believe you should be able to use the `OverridableIndex` as is in Jandex. See https://github.com/quarkusio/quarkus/blob/db7afb1ac384d17eda19c81c54505432438c4068/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/OverridableIndex.java

If you think this is ok, please review the PR. Thanks

Reply:
I don't think your `annotationInstancenfoComparator` does what you want it to do. If I understand it correctly, the `getAnnotations` method would only ever return _one_ instance of any given annotation. Annotations would have to be unified across annotation targets, not globally.

The other methods seem fine.

Reply:
@Ladicek - Thanks ! So I am correct in saying that if I use name and name of the target then above will be correct ?

Reply:
Comparing annotation targets is gonna be ... difficult. For `ClassInfo`, comparing the name should be enough. For `FieldInfo`, comparing the class name and the field name. For `MethodInfo`, comparing the class name, the method name and the parameter list. For type targets... I don't really know off hand.

Reply:
OK thanks . Let me give it a shot. I'll ping you once it's updated

Reply:
@Ladicek - please have a look again. This definitely needs to be in Jandex rather than here, but for now this works for my use case, if and when you move this to Jandex, I can remove this here.

Reply:
I'm not exactly sure about comparing types by names (because Jandex returns `Number` as the name of `? extends Number`, so `Number` and `? extends Number` would be equal), but what you have now should work in 99.99% cases, so it's probably fine right now :-)

Reply:
Ok thanks, please let me know if you are pulling this into Jandex, so then I can remove it here once available. Thanks for your help

Reply:
I'll add something to Jandex for sure (though personally, I see this as a Quarkus design problem and creating ad hoc overlay indices is just a band-aid).